python.d/boinc.chart.py

```diff
@@ -0,0 +1,170 @@
+        ]
+    },
+    'process_states': {
+        'options': [None, 'Tasks per Process State', 'tasks', 'boinc', 'boinc.process', 'line'],
+        'lines': [
+            ['uninit_proc', 'Uninitialized', 'absolute', 1, 1],
+            ['executing', 'Executing', 'absolute', 1, 1],
+            ['suspended', 'Suspended', 'absolute', 1, 1],
+            ['aborting', 'Aborted', 'absolute', 1, 1],
+            ['quit', 'Quit', 'absolute', 1, 1],
+            ['copy_pending', 'Copy Pending', 'absolute', 1, 1]
+        ]
+    }
+}
+
+# A simple template used for pre-loading the return dictionary to make
+# the _get_data() method simpler.
+_DATA_TEMPLATE = {
+    'total': 0,
+    'active': 0,
+    'new': 0,
+    'downloading': 0,
+    'downloaded': 0,
+    'comperror': 0,
+    'uploading': 0,
+    'uploaded': 0,
+    'aborted': 0,
+    'upload_failed': 0,
+    'uninit_sched': 0,
+    'preempted': 0,
+    'scheduled': 0,
+    'uninit_proc': 0,
+    'executing': 0,
+    'suspended': 0,
+    'aborting': 0,
+    'quit': 0,
+    'copy_pending': 0
+}
+
+# Map task states to dimensions
+_TASK_MAP = {
+    boinc_client.ResultState.NEW: 'new',
+    boinc_client.ResultState.FILES_DOWNLOADING: 'downloading',
+    boinc_client.ResultState.FILES_DOWNLOADED: 'downloaded',
+    boinc_client.ResultState.COMPUTE_ERROR: 'comperror',
+    boinc_client.ResultState.FILES_UPLOADING: 'uploading',
+    boinc_client.ResultState.FILES_UPLOADED: 'uploaded',
+    boinc_client.ResultState.ABORTED: 'aborted',
+    boinc_client.ResultState.UPLOAD_FAILED: 'upload_failed'
+}
+
+# Map scheduler states to dimensions
+_SCHED_MAP = {
+    boinc_client.CpuSched.UNINITIALIZED: 'uninit_sched',
+    boinc_client.CpuSched.PREEMPTED: 'preempted',
+    boinc_client.CpuSched.SCHEDULED: 'scheduled',
+}
+
+# Maps process states to dimensions
+_PROC_MAP = {
+    boinc_client.Process.UNINITIALIZED: 'uninit_proc',
+    boinc_client.Process.EXECUTING: 'executing',
+    boinc_client.Process.SUSPENDED: 'suspended',
+    boinc_client.Process.ABORT_PENDING: 'aborted',
+    boinc_client.Process.QUIT_PENDING: 'quit',
+    boinc_client.Process.COPY_PENDING: 'copy_pending'
+}
+
+
+class Service(SimpleService):
+    def __init__(self, configuration=None, name=None):
+        SimpleService.__init__(self, configuration=configuration, name=name)
+        self.order = ORDER
+        self.definitions = CHARTS
+        self.host = self.configuration.get('host', 'localhost')
+        self.port = self.configuration.get('port', 0)
+        self.password = self.configuration.get('password', '')
+        self.client = boinc_client.BoincClient(host=self.host, port=self.port, passwd=self.password)
+        self.alive = False
+
+    def check(self):
+        if platform.system() != 'Linux':
+            self.error('Only supported on Linux.')
+            return False
+        self.connect()
+        self.alive = self.client.connected and self.client.authorized:
+        return self.alive
```

Comment:
code right after platform check looks like the same as `self.reconnect` (`connect; set alive; return alive`)
i think you can `return self.reconnect()` after platform check.

(in this case `self.connect` method will not be used anywhere)

what do you think?

Reply:
Functionally, they're equivalent.  I'm not a big fan of running code in try blocks that I know will fail though, and the way it's currently written is consistent with the layout used for this type of thing in other plugins (both spigotmc and beanstalkd do things this way).  If it should be changed, I'd almost rather:

```python
    def check(self):
        if platform.system() != 'Linux':
            self.error('Only supported on Linux.')
            return False
        return self.connect()

    def connect(self):
        self.client.connect()
        self.alive = self.client.connected and self.client.authorized:
        return self.alive

    def reconnect(self):
        try:
            self.client.disconnect()
        except socket.error:
            pass
        return self.connect()
```

That still minimizes the amount of code, while providing the same set of methods that the other modules do, and avoiding trying to run `self.client.disconnect()` during `check()`.

Reply:
Nice :+1: i like it

Reply:
OK, I'll put together a commit that updates it to use that style.